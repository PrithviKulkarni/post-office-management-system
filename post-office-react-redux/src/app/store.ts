import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import letterReducer from '../features/LetterSlice';
import parcelReducer from '../features/ParcelSlice';
import depotReducer from '../features/DepotSlice';
import driverReducer from '../features/DriverSlice' 
export const store = configureStore({
  reducer: {
    letter: letterReducer,
    parcel: parcelReducer,
    depot: depotReducer,
    driver: driverReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
