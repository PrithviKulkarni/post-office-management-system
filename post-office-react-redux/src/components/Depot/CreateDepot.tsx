import React, { useState, SyntheticEvent } from 'react'

import { useNavigate } from 'react-router-dom'

import { useAppDispatch } from '../../app/hooks'

import { Container, Button, Form, InputGroup } from 'react-bootstrap'
import { Depot, defaultDepot } from '../interfaces/depotinterface'

import { addNewDepot } from '../../features/DepotSlice'

const CreateDepot = () => {
  const navigate = useNavigate()
  const dispatch = useAppDispatch()

  const [city, setCity] = useState<string>('')
  const [errorMessage, setErrorMessage] = useState<string>('')
  const [workers, setWorkers] = useState<number>(0)

  const submitHandler = (event: SyntheticEvent) => {
    event.preventDefault()

    const validated = [city, workers].every(Boolean)
    try {
      if (validated) {
        const newDepotData: Depot = defaultDepot

        if (city) {
          newDepotData.city = city
        }
        if (workers) {
          newDepotData.workers = Number(workers)
        }

        console.log({ newDepotData })

        dispatch(addNewDepot(newDepotData))
        navigate('/depots')
      }
    } catch (error) {
      setErrorMessage('Unable to create depot.')
      console.log('errorMessage: ', errorMessage)
    }
  }

  return (
    <div className="createDepot" data-testid='createDepot'>
      <Container>
        <Form onSubmit={submitHandler}>
          <InputGroup className='mb-3'>
            <InputGroup.Text id='basic-addon1'>Depot City</InputGroup.Text>
            <Form.Control
              type='text'
              aria-describedby='basic-addon1'
              size='sm'
              data-testid="cityIn"
              value={city}
              onChange={event => setCity(event.target.value)}
            />
          </InputGroup>

          <InputGroup className='mb-3'>
            <InputGroup.Text id='basic-addon1'>
              Number of workers
            </InputGroup.Text>
            <Form.Control
              type='text'
              aria-describedby='basic-addon1'
              size='sm'
              data-testid="workersIn"
              value={workers}
              onChange={event => setWorkers(Number(event.target.value))}
            />
          </InputGroup>

          <Button type='submit' className="button" data-testid="button">Create Depot</Button>
          <p>{errorMessage}</p>
        </Form>
      </Container>
    </div>
  )
}

export default CreateDepot
