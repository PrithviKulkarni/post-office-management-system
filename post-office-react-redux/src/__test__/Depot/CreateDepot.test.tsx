import { render, cleanup, fireEvent } from '@testing-library/react';
import CreateDepot from '../../components/Depot/CreateDepot';
import { screen } from '@testing-library/react';
import { Provider } from "react-redux";
import { store } from '../../app/store';
import { BrowserRouter } from 'react-router-dom';

afterEach(cleanup);
it("Depot create page render test", () => {
  const { asFragment } = render(
    <Provider store={store}>
      <BrowserRouter>
      <CreateDepot /></BrowserRouter>
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("Verifies depot create has rendered", () => {
  const { getByTestId } = render(
    <Provider store={store}>
      <BrowserRouter>
      <CreateDepot />
      </BrowserRouter>
    </Provider>
  );
  const button = getByTestId("button");
  expect(button.textContent).toBe("Create Depot");
  const city = getByTestId("cityIn");
  const workers = getByTestId("workersIn");
  fireEvent.change(city, { target: { value: "test" } });
  fireEvent.change(workers, { target: { value: "27" } });
  fireEvent.click(button);
  expect(button).toBeTruthy();
  expect(screen.getByTestId("createDepot")).toHaveClass("createDepot");
});

