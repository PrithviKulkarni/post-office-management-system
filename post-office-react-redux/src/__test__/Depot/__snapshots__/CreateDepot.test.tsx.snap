// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Depot create page render test 1`] = `
<DocumentFragment>
  <div
    class="createDepot"
    data-testid="createDepot"
  >
    <div
      class="container"
    >
      <form
        class=""
      >
        <div
          class="mb-3 input-group"
        >
          <span
            class="input-group-text"
            id="basic-addon1"
          >
            Depot City
          </span>
          <input
            aria-describedby="basic-addon1"
            class="form-control form-control-sm"
            data-testid="cityIn"
            type="text"
            value=""
          />
        </div>
        <div
          class="mb-3 input-group"
        >
          <span
            class="input-group-text"
            id="basic-addon1"
          >
            Number of workers
          </span>
          <input
            aria-describedby="basic-addon1"
            class="form-control form-control-sm"
            data-testid="workersIn"
            type="text"
            value="0"
          />
        </div>
        <button
          class="button btn btn-primary"
          data-testid="button"
          type="submit"
        >
          Create Depot
        </button>
        <p />
      </form>
    </div>
  </div>
</DocumentFragment>
`;
